### PR TITLE
Add LoRA LR multiplier ablation and fix instruct training pipeline

### LINKED ISSUE #19: Ablate Differential LoRA Learning Rate Alongside Merge Ratio Sweeps
Investigate the effect of differential learning rates for LoRA matrices A and B during multimodal adapter training. Maya found that using equal learning rates for LoRA A and B at 8B scale led to LoRA failure, while Aya Vision used differential rates successfully. Whether this failure mode reproduces at 3.35B is an open question.

This ablation should run early in Phase 2 alongside the merge ratio sweeps (Issue #18 ) so that findings can inform the remaining training runs. At minimum, compare equal vs. differential LoRA learning rates (e.g., lr_A != lr_B following Aya Vision's configuration) and observe the effect on training stability and downstream performance.

### Context

- Maya (arXiv:2412.07112) reported LoRA instability at 8B with equal A/B learning rates.
- Aya Vision applied differential LoRA learning rates as part of their training recipe.
- At 3.35B, LoRA dynamics may differ due to the smaller model having different parameter redundancy characteristics.
- This directly feeds into the Phase 3 LoRA rank sweep (Gap 1).

### Dependencies

- Phase 1 infrastructure and training pipeline.
- Should run concurrently with or immediately before Issue #18 (merge ratio ablation), since the best LoRA configuration should be used for the merge sweeps.

## Acceptance Criteria

- [ ] At least two LoRA learning rate configurations compared: (1) equal A/B rates, (2) differential A/B rates (following Aya Vision's recipe or a principled variant).
- [ ] Training stability documented: loss curves, gradient norms, and any divergence/instability observed under equal rates.
- [ ] Both configurations evaluated on at least one benchmark to assess downstream impact.
- [ ] A clear recommendation is recorded for which LoRA LR configuration to use in subsequent Phase 2 and Phase 3 training.
- [ ] Results and rationale documented in a brief write-up or experiment log.

## Estimated Effort

1--2 days (2 training runs + comparison)

### PR DESCRIPTION
**Pull request by engichang1467**
## Summary

This PR sets up the infrastructure to run the differential LoRA learning rate ablation described in #19, and fixes several issues that were blocking the instruction fine-tuning pipeline from running on Modal.

## Changes

- **LoRA LR multiplier support** — `modal_train_instruct.py` now accepts `--lora-a-lr-multiplier` and `--lora-b-lr-multiplier` CLI flags, which are forwarded through to `LoraAdapterConfig`. This makes it trivial to launch side-by-side runs comparing equal A/B rates (both `1.0`) vs. a differential schedule (e.g., the Aya Vision recipe), directly addressing the ablation criteria in #19.

- **Dataset switch to LLaVA-Instruct-150K** — `instruct.yaml` was pointing at the full 665K mix; switched to `llava_instruct_150k` for the ablation runs to keep GPU time manageable.

- **Training hyperparameter fixes** — Corrected `InstructConfig` defaults: `batch_size` reduced to `4` (per-GPU) with `grad_acc_steps=128` to maintain an effective batch size of 512; added an explicit `max_seq_len=1536` since instruct conversations are longer than pretrain captions.

- **Gradient checkpointing re-enabled** — The `gradient_checkpointing_enable()` call in `train_instruct.py` was commented out; re-enabled it (non-reentrant) and fixed the path to use `language_model` directly (no `base_model` wrapper at this point). Also switched `torch.compile` to `dynamic=False` for compatibility.

- **`modal_download.py` overhaul** — Expanded to support downloading both the pretrain and instruct datasets (`--dataset pretrain|instruct|all`). Instruct images come from COCO train2017 (no HF zip); both datasets are now stored as unextracted zips in the volume to stay under Modal's 500K inode limit. Extraction happens at training time inside the Modal function. Added `hf_transfer` for faster downloads.

- **`modal_train_instruct.py` fixes** — Upgraded default GPU from A10G → A100-40GB, added `hydra-core` to the Modal image, wired in on-demand COCO zip extraction to `/tmp/llava-instruct` at the start of the training function, and fixed `data_dir` to point to the extraction target.

- **Image path resolution fix in `data.py`** — `InstructDataset` now uses `_resolve_image_path()` instead of a bare `data_dir / image` join, fixing filtering for the COCO subdirectory layout.

## How this resolves #19

The two acceptance criteria runs can now be kicked off as:

```bash
# Equal LR (baseline)
modal run --detach scripts/modal_train_instruct.py --alignment-checkpoint /models/<run_id>/checkpoint_<step>.pt

# Differential LR (Aya Vision recipe)
modal run --detach scripts/modal_train_instruct.py \
  --alignment-checkpoint /models/<run_id>/checkpoint_<step>.pt \
  --lora-a-lr-multiplier <A_mult> --lora-b-lr-multiplier <B_mult>
```

## Test plan

- [ ] `pytest tests/ -v` passes locally
- [ ] `modal run scripts/modal_download.py --dataset instruct` downloads and commits the COCO zip without inode errors
- [ ] Instruct training run launches on Modal without import/config errors
- [ ] Baseline (equal LR) and differential-LR ablation runs both produce stable loss curves for comparison

### CODE REVIEW

**Comment by sanggusti** [comment]
> /run scripts/modal_download.py --dataset instruct

https://github.com/Cohere-Labs-Community/expedition-tayavision/actions/runs/24818856041

seems like some error in the parsing in `ci-cd.yaml`. I didn't program it to take argv input.

**Comment by sanggusti** [comment]
Error: Invalid command. Usage: `/run scripts/modal_<name>.py`
Example: `/run scripts/modal_train_alignment.py`

**Comment by sanggusti** [comment]
image
I've added the instruct datasets. When running the modal via ci-cd try not adding arguments. So it would be just like "/run scripts/modal_xxx.py"

**Review by sanggusti — APPROVED** [review]

**Comment by sanggusti** [comment]
image
what a luck, it only runs for 6 hours because github actions have time limit of 6 hours. But the training loss looks promising
image
https://wandb.ai/gustiwinata/tayavision-instruct

**Comment by sanggusti** [comment]
Need to add checkpoint saving for every progress. Currently by modal it suddenly stopped in 60% and not find any latest checkpoint making it starting from scratch.

**Comment by engichang1467** [comment]
# Update

## Problem

@sanggusti and @engichang1467  run on Modal keeps dying midway with no recovery. The root cause is that GPU containers on Modal are always preemptible (spot instances), and there was no protection against it.

## Change 1 — `config/training_config.py`

```
save_steps: int = 50
```
**The bug**: `InstructConfig` inherited `save_steps = 500` from `AlignmentConfig`, but instruct training only produces ~292 total optimizer steps (150K samples ÷ batch_size=4 ÷ grad_acc_steps=128). `500 > 292`, so the checkpoint-saving condition in the training loop was never triggered. The only checkpoint ever written was the one at the very end of training. Any preemption before that meant starting over completely.

**The fix**: Override `save_steps = 50` in `InstructConfig`, giving ~6 checkpoints spread across the run. It also protects against other failure modes (OOM, crashes) beyond just preemption.

## Change 2 — `pipeline/train_instruct.py`

```python
current_step = [step_offset]
if is_main:
    def _sigterm_handler(signum, frame):
        tqdm.write(f"SIGTERM received — saving emergency checkpoint at step {current_step[0]}")
        save_checkpoint(checkpoint_dir, current_step[0], model, optimizer, lr_scheduler)
        sys.exit(0)
    signal.signal(signal.SIGTERM, _sigterm_handler)
```

**Why**: Modal sends `SIGTERM` to the container before evicting it, with a short grace period. Without a handler, Python's default response to SIGTERM is immediate termination — no checkpoint, no cleanup. This handler intercepts the signal and saves a checkpoint from the exact step that was last completed. `current_step` is a single-element list (rather than a plain int) so the nested handler function can update it via closure as training progresses.

Change 3 — `scripts/modal_train_instruct.py`

```python
try:
    main(...)
finally:
    models_volume.commit()
```

**Why**: This is what makes Change 2 actually work. Modal Volumes don't guarantee that writes survive an abnormal container exit unless `volume.commit()` is called. The SIGTERM handler calls `sys.exit(0)` which raises `SystemExit` — Python's `finally` blocks still run on `SystemExit`, so `models_volume.commit()` fires before the process dies, flushing the just-written checkpoint to Modal's persistent storage. Without this, the file might be written to the container's local buffer but lost when the container is reclaimed.

## End-to-end recovery flow

When preemption happens:

1. Modal sends `SIGTERM`
2. Handler saves checkpoint to `/models/<run_id>/checkpoint_<step>.pt`
3. `sys.exit(0)` → `finally` → `models_volume.commit()` persists it
4. Your friend re-runs with `--resume-run-id <id>` and picks up from the last saved step

**Comment by sanggusti** [comment]
running it right now

**Comment by sanggusti** [comment]
image
Still didn't save any checkpoint after 40%

```

Runner interrupted due to worker preemption. Your Function will be restarted with the same input. For more details, see https://modal.com/docs/guide/preemption
--
  | Extracting /data/llava-instruct/coco_train2017.zip to /tmp/llava-instruct ...
  | Extraction complete.
  | Single-GPU: world_size=1, global_batch_size=4, per_gpu_batch_size=4
  | Run ID: a6d61ed8-54bd-4d76-831c-24b279bbd77c
  | Checkpoint dir: /models/a6d61ed8-54bd-4d76-831c-24b279bbd77c


```

**Comment by Rahul007007** [comment]
Here are the ablated models:
1. https://huggingface.co/TrishanuDas/tayavision-inst-sweep-lrm_a_1_lrm_b_1
2. https://huggingface.co/TrishanuDas/tayavision-inst-sweep-lrm_a_1_lrm_b_2
3. https://huggingface.co/TrishanuDas/tayavision-inst-sweep-lrm_a_2_lrm_b_1
4. https://huggingface.co/TrishanuDas/tayavision-inst-sweep-lrm_a_1_lrm_b_0.1

WandB report: https://wandb.ai/rahul007-/tayavision-instruct-sweep/reports/LoRA-learning-rate-ablations--VmlldzoxNjY4ODMwMQ

**Comment by sanggusti** [comment]
Just resolved the conflict, please rerun
